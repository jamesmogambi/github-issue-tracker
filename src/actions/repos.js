require("isomorphic-fetch");

// SET_REPOSITORIES
export const setRepositories = (payload) => ({
  type: "SET_REPOSITORIES",
  payload,
});
// CLEAR_REPOSITORIES
export const clearRepositories = () => ({
  type: "CLEAR_REPOSITORIES",
});

// ERROR
export const error = (payload) => ({
  type: "ERROR",
  payload,
});

// Get Github Repositories
export const getRepositories = (params) => {
  return (dispatch) => {
    dispatch(clearRepositories());
    fetch("https://api.github.com/graphql", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        authorization: `bearer ${process.env.GITHUB_ACCESS_TOKEN}`,
      },
      body: JSON.stringify({
        query: `
               query($github_login: String!) {
    user(login: $github_login) {
      name
      repositories(first: 100) { 
        nodes {
             id
        name
        description 
        url 
        updatedAt
        forkCount
        }
      }
    }
  }
                 `,
        variables: {
          github_login: params.username,
        },
      }),
    })
      .then((res) => res.json())
      .then((res) => {
        if (res.data.user) {
          let repos = res.data.user.repositories.nodes;
          let owner = res.data.user.name;
          dispatch(
            setRepositories({
              repositories: repos,
              repositoryOwner: params.username,
            })
          );
        }
      })
      .catch((err) => {
        dispatch(error(err));
        console.error(err);
      });
  };
};
