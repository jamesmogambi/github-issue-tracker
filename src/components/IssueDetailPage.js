import React from "react";
import { connect } from "react-redux";

export const IssueDetailsPage = (props) => (
  <div>
    <div className="content-container">
      {props.issue ? (
        <div>
          <h1 className="page-header__title">
            Viewing details of Issue <span>{props.issue.node.title}</span>{" "}
          </h1>
          <div
            dangerouslySetInnerHTML={{ __html: props.issue.node.bodyHTML }}
          ></div>
        </div>
      ) : (
        <div className="list-item list-item--message">
          <span>No Details </span>
        </div>
      )}
    </div>
  </div>
);

const mapStateToProps = (state, props) => {
  return {
    issue: state.issues.repositoryissues.find(
      (issue) => issue.node.id === props.match.params.id
    ),
  };
};

export default connect(mapStateToProps)(IssueDetailsPage);
