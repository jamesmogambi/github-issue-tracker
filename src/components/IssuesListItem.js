import React from "react";
import { Link } from "react-router-dom";

const IssuesListItem = ({ id, title, author, createdAt }) => (
  <Link className="list-item" to={`/issueDetails/${id}`}>
    <div>
      <img src={author.avatarUrl} alt="Avatar" className="avatar" />
      <div className="list-item__sub-title">{title}</div>
    </div>
    <span className="list-item__sub-title ">opened at {createdAt}</span>
  </Link>
);

export default IssuesListItem;
