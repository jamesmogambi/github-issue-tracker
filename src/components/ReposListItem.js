import React from "react";
import { Link } from "react-router-dom";

const ReposListItem = ({ id, name, description, repositoryOwner }) => (
  <div className="list-item">
    <div> 
      <h3 className=" list-item__title">{name}</h3>
      <span className=" list-item__sub-title">{description}</span>
      <Link to={`/issues/${name}/${repositoryOwner}`}>
        <h3 className="list-item__sub-title list-link ">View Issues</h3>
      </Link>
    </div>
    <a className="button button--link " href={`https://github.com/${repositoryOwner}/${name}/tags`} target="_blank">Tags</a>
 </div>
);

export default ReposListItem;
