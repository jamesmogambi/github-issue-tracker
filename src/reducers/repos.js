// Repositories Reducer
const initialState = {
  repositories: [],
  repositoryOwner: null,
  error: null,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case "SET_REPOSITORIES":
      const { repositories, repositoryOwner } = action.payload;
      return {
        ...state,
        repositoryOwner,
        repositories,
      };
    case "CLEAR_REPOSITORIES":
      return {
        repositories: [],
        repositoryOwner: null,
        error: null,
      };

    default:
      return state;
  }
};
