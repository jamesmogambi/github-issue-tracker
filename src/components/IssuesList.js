import React from "react";
import { connect } from "react-redux";
import { getRepositoryIssues } from "../actions/issues";
import IssuesListItem from "./IssuesListItem";
import selectIssues from "../selectors/issues";

// export class IssuesList extends React.Component {
//   constructor(props) {
//     super(props);
//   }
//   componentWillMount() {
//     this.props.getRepositoryIssues(this.props.params);
//   }
//   render() {
//     return (
//       <div className="content-container">
//         <h1 className="page-header__title">
//           Viewing Issues of Repository <span>{this.props.params.repoName}</span>{" "}
//         </h1>

//         <div className="list-header">
//           <div className="show-for-mobile">Issues</div>
//           <div className="show-for-desktop">Issue</div>
//           <div className="show-for-desktop">Created</div>
//         </div>
//         <div className="list-body">
//           {this.props.issues.length === 0 ? (
//             <div className="list-item list-item--message">
//               <span>No Issues for this Repo</span>
//             </div>
//           ) : (
//             this.props.issues.map(({ node }) => {
//               return <IssuesListItem key={node.id} {...node} />;
//             })
//           )}
//         </div>
//       </div>
//     );
//   }
// }

// const mapDispatchToProps = (dispatch) => ({
//   getRepositoryIssues: (params) => dispatch(getRepositoryIssues(params)),
// });

// const mapStateToProps = (state) => {
//   return {
//     issues: selectIssues(state.issues.repositoryissues, state.filters),
//   };
// };

// export default connect(mapStateToProps, mapDispatchToProps)(IssuesList);

export const IssuesList = (props) => (
  <div className="content-container">
    <h1 className="page-header__title">
      Viewing Issues of Repository <span>{props.repoName}</span>{" "}
    </h1>

    <div className="list-header">
      <div className="show-for-mobile">Issues</div>
      <div className="show-for-desktop">Issue</div>
      <div className="show-for-desktop">Created</div>
    </div>
    <div className="list-body">
      {props.issues.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No Issues for this Repo</span>
        </div>
      ) : (
        props.issues.map(({ node }) => {
          return <IssuesListItem key={node.id} {...node} />;
        })
      )}
    </div>
  </div>
);

export default IssuesList;
