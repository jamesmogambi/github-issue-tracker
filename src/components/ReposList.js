import React from "react";
import { connect } from "react-redux";
import ReposListItem from "./ReposListItem";

export const ReposList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Repositories</div>
      <div className="show-for-desktop">Repository</div>
      <div className="show-for-desktop">Action</div>
    </div>
    <div className="list-body">
      {props.repos.repositories.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No Repos</span>
        </div>
      ) : (
        props.repos.repositories.map((repo) => {
          return (
            <ReposListItem
              key={repo.id}
              repositoryOwner={props.repos.repositoryOwner}
              {...repo}
            />
          );
        })
      )}
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    repos: state.repos,
  };
};

export default connect(mapStateToProps)(ReposList);
