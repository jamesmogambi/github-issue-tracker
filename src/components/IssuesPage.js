import React from "react";
import { connect } from "react-redux";
import { getRepositoryIssues } from "../actions/issues";
import IssuesList from "./IssuesList";
import IssuesListFilters from "./IssuesListFilters";
import selectIssues from "../selectors/issues";

export class IssuesPage extends React.Component {
  constructor(props) {
    super(props);
  }
  componentWillMount() {
    this.props.getRepositoryIssues(this.props.match.params);
  }
  render() {
    return (
      <div>
        <IssuesListFilters />
        <IssuesList issues={this.props.issues} repoName={this.props.repoName} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  getRepositoryIssues: (params) => dispatch(getRepositoryIssues(params)),
});

const mapStateToProps = (state) => {
  return {
    issues: selectIssues(state.issues.repositoryissues, state.filters),
    repoName: state.issues.repositoryname,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(IssuesPage);
