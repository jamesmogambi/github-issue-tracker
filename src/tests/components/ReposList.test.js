import React from "react";
import { shallow } from "enzyme";
import { ReposList } from "../../components/ReposList";
import repos from "../fixtures/repos";

test("should render ReposList with repositories", () => {
  const data = {
    repositoryOwner: "JamesEdward",
    repositories: repos,
  };
  const wrapper = shallow(<ReposList repos={data} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ReposList with empty message", () => {
  const data = {
    repositoryOwner: null,
    repositories: [],
  };
  const wrapper = shallow(<ReposList repos={data} />);
  expect(wrapper).toMatchSnapshot();
});
