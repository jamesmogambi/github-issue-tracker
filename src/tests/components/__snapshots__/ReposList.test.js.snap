// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ReposList with empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Repositories
    </div>
    <div
      className="show-for-desktop"
    >
      Repository
    </div>
    <div
      className="show-for-desktop"
    >
      Action
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No Repos
      </span>
    </div>
  </div>
</div>
`;

exports[`should render ReposList with repositories 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Repositories
    </div>
    <div
      className="show-for-desktop"
    >
      Repository
    </div>
    <div
      className="show-for-desktop"
    >
      Action
    </div>
  </div>
  <div
    className="list-body"
  >
    <ReposListItem
      description="view awesome places."
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkxNjk1NjQ3NTA="
      name="react-native-awesome-places-app"
      repositoryOwner="JamesEdward"
      updatedAt="2019-06-05T10:58:19Z"
      url="https://github.com/jamesmogambi/react-native-awesome-places-app"
    />
    <ReposListItem
      description={null}
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkxNzAxOTY3OTU="
      name="react-indecision-app-v"
      repositoryOwner="JamesEdward"
      updatedAt="2019-06-05T10:56:56Z"
      url="https://github.com/jamesmogambi/react-indecision-app-v"
    />
    <ReposListItem
      description="a frontend-development test app"
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkxOTcwNjY4MTI="
      name="react-interview-app"
      repositoryOwner="JamesEdward"
      updatedAt="2019-07-15T21:27:52Z"
      url="https://github.com/jamesmogambi/react-interview-app"
    />
    <ReposListItem
      description="emr solutions app"
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkxOTkyODE2MTc="
      name="medescope"
      repositoryOwner="JamesEdward"
      updatedAt="2020-12-16T20:34:45Z"
      url="https://github.com/jamesmogambi/medescope"
    />
    <ReposListItem
      description="baobab-app"
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkyMjIxMTQ2NTY="
      name="baobab-app"
      repositoryOwner="JamesEdward"
      updatedAt="2019-11-16T15:01:01Z"
      url="https://github.com/jamesmogambi/baobab-app"
    />
    <ReposListItem
      description={null}
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkyNjAxNTIxMTA="
      name="inventory"
      repositoryOwner="JamesEdward"
      updatedAt="2020-04-30T08:15:53Z"
      url="https://github.com/jamesmogambi/inventory"
    />
    <ReposListItem
      description={null}
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkyNjAyNTM4ODQ="
      name="inventorygrain"
      repositoryOwner="JamesEdward"
      updatedAt="2020-04-30T15:55:34Z"
      url="https://github.com/jamesmogambi/inventorygrain"
    />
    <ReposListItem
      description="Leave Management System for CSI with a Spring Boot backend and a React frontend."
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkyNjE2ODg0OTE="
      name="csileavemanagementsystem"
      repositoryOwner="JamesEdward"
      updatedAt="2020-05-06T07:50:11Z"
      url="https://github.com/jamesmogambi/csileavemanagementsystem"
    />
    <ReposListItem
      description="Application for finding vets/animal health specialists,booking appointments."
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkyOTAxMzEzNzA="
      name="team-alpha-app"
      repositoryOwner="JamesEdward"
      updatedAt="2020-08-26T09:07:10Z"
      url="https://github.com/teamAlpha693/team-alpha-app"
    />
    <ReposListItem
      description="Django backend for eCommerce project"
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkzMTU5NDY3NTM="
      name="ecommerce-backend"
      repositoryOwner="JamesEdward"
      updatedAt="2020-11-25T13:24:51Z"
      url="https://github.com/jamesmogambi/ecommerce-backend"
    />
    <ReposListItem
      description="Fullstack ecommerce application built with React, Node, Express and MongoDB."
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkzMTYwMDIzMjA="
      name="MERN-Ecommerce"
      repositoryOwner="JamesEdward"
      updatedAt="2021-01-28T10:34:46Z"
      url="https://github.com/jamesmogambi/MERN-Ecommerce"
    />
    <ReposListItem
      description="Social network for developers, built on the MERN stack"
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkzMTk1NzkzNDQ="
      name="dev-_MERN-Stack"
      repositoryOwner="JamesEdward"
      updatedAt="2021-01-28T10:36:27Z"
      url="https://github.com/jamesmogambi/dev-_MERN-Stack"
    />
    <ReposListItem
      description="Manage your expenses."
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkzMjExNTE3Nzk="
      name="react-redux-firebase-expensify-app"
      repositoryOwner="JamesEdward"
      updatedAt="2021-03-17T14:38:54Z"
      url="https://github.com/jamesmogambi/react-redux-firebase-expensify-app"
    />
    <ReposListItem
      description="Payment App UI developed in React Native."
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkzMzE1ODU2NDg="
      name="fintech-front-end"
      repositoryOwner="JamesEdward"
      updatedAt="2021-02-20T08:04:33Z"
      url="https://github.com/jamesmogambi/fintech-front-end"
    />
    <ReposListItem
      description={null}
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkzMzIzOTE0NTc="
      name="0nline-payment-admin"
      repositoryOwner="JamesEdward"
      updatedAt="2021-01-24T07:29:45Z"
      url="https://github.com/jamesmogambi/0nline-payment-admin"
    />
    <ReposListItem
      description="Progressive web app for viewing issues from the github.com. Built with react, redux, redux-thunk and webpack."
      forkCount={0}
      id="MDEwOlJlcG9zaXRvcnkzNTAyODE0NTA="
      name="github-issue-tracker"
      repositoryOwner="JamesEdward"
      updatedAt="2021-03-22T14:02:44Z"
      url="https://github.com/jamesmogambi/github-issue-tracker"
    />
  </div>
</div>
`;
